### ISSUE TITLE
Eclipse Product Export Wizard fails silently, error only shown in log without information

### ISSUE THREAD

**Issue by thahnen** [eclipse-pde/eclipse.pde#825]
# TL;DR

Started working with Eclipse rich client applications based on [THIS](https://www.vogella.com/tutorials/EclipseRCP/article.html) tutorial. I set everything step-by-step. When trying to use the **Eclipse Product Export Wizard** (clicking Finish after doing all the configuration) from the product overview page, nothing happened but when looking at the _Error Log_ view, this `Unhandled event loop exception` showed up:

`java.lang.IllegalStateException: Exporting mixed products is not (yet) supported
	at org.eclipse.pde.internal.ui.wizards.exports.ProductExportWizard.scheduleExportJob(ProductExportWizard.java:92)
	at org.eclipse.pde.internal.ui.wizards.exports.BaseExportWizard.performFinish(BaseExportWizard.java:79)
	at org.eclipse.jface.wizard.WizardDialog.finishPressed(WizardDialog.java:837)
	at org.eclipse.jface.wizard.WizardDialog.buttonPressed(WizardDialog.java:473)
	at org.eclipse.jface.dialogs.Dialog.lambda$0(Dialog.java:619)
	at org.eclipse.swt.events.SelectionListener$1.widgetSelected(SelectionListener.java:84)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4646)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1524)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1547)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1532)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1325)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4413)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3989)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.pde.internal.ui.editor.product.ProductExportAction.run(ProductExportAction.java:57)
	at org.eclipse.pde.internal.ui.editor.product.OverviewPage.linkActivated(OverviewPage.java:107)
	at org.eclipse.ui.forms.widgets.FormText.activateLink(FormText.java:1544)
	at org.eclipse.ui.forms.widgets.FormText.handleMouseClick(FormText.java:1355)
	at org.eclipse.ui.forms.widgets.FormText$2.mouseUp(FormText.java:440)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:224)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4646)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1524)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1547)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1532)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1325)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4413)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3989)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:342)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:645)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:342)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:552)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:171)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:651)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:588)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1459)`

Looking at Google or old issues did not give a precise hint on what went wrong (the stack trace is not very much explanatory or links to further documentation). What caused this error to happen? It does not seem like a bug but intended behaviour, so what is wrong with the plugin / product configuration?

Additionally, it would be really nice to get this kind of feedback directly when clicking on Finish in the wizard, having the _Error Log_ view open was just coincidence.

## Environment

I'm working with Eclipse RCP 2023-09 on an Intel macOS 13.6 device.

**Comment by HannesWell** [comment]
> Looking at Google or old issues did not give a precise hint on what went wrong (the stack trace is not very much explanatory or links to further documentation). What caused this error to happen? It does not seem like a bug but intended behaviour, so what is wrong with the plugin / product configuration?

Yes mixed products are a relativly new Feature, therefore there are probably not many references in the Internet

The issue is, that exporting Mixed products from the workspace is simply not yet implemented. It hopefully is one day in the future, but at the Moment you can only build them with Maven+Tycho.

@vogella I think this has to be mentioned in your Tutorial. :/

If you only have Features in your product you can simply Change the Type ('product is based on') from mixed (`Plugins and Features`) to only `Features`.

> Additionally, it would be really nice to get this kind of feedback directly when clicking on Finish in the wizard, having the Error Log view open was just coincidence.

Agree. It would probably best to even disable the export Button for mixed Products with a Tool Tip indicating what's happening.

**Comment by thahnen** [comment]
So the workaround in this case would be to have a feature containing the plugin and then change the product based on type to feature and reference this one? If so, I'll use this workaround but TBH it sounds like what I want to do with extra steps :smile:

**Comment by laeubi** [comment]
@thahnen the IDE export is based on pdebuild and this project is almost dead see 

https://github.com/eclipse-pde/eclipse.pde/tree/master/build

so the main question is, why you need to export it from the IDE at all? Setup a pomless Tycho build is rather trivial and will give you the most flexible, powerful and supported to the bleeding edge approach.

**Comment by thahnen** [comment]
Wasn't aware that the project is dead / the direct replacement is to use Tycho. I was thinking about setting up a Maven build including Tycho anyway in the future when my project evolves from "me-tinkering-late-at-night" to something more stable.

After testing further with the **Eclipse Product Export Wizard**, I also noticed that on macOS the exported "application" is not really an application bundle and fails to start. But it also doesn't warn before exporting that some further configuration from my side is missing to get it up and running.

Long story short, I'll directly move to Tycho and ignore the wizard completely (as I planned to do so later on anyway). I conclude this issue thread with only some "feature requests":

- Move the feedback from the Error Log to the **Eclipse Product Export Wizard** or disable the export button (I'd prefer the second option regarding UX)
- Somehow warn in the future when using any kind of IDE export (I assume the export of plug-ins, etc. is therefore deprecated as well) and notify the user about the replacement using Tycho
- Nice-to-have: Update the tutorial @vogella 😄

**Comment by vogella** [comment]
Please point me to the place in which my tutorials still advocates for using the pde export wizard. I can update it, but I thought I did a long time ago

**Comment by HannesWell** [comment]
> So the workaround in this case would be to have a feature containing the plugin and then change the product based on type to feature and reference this one? If so, I'll use this workaround but TBH it sounds like what I want to do with extra steps 😄

That's totally right and the reason why Mixed products were introduced recently. They are simply not (yet) supported by the workspace Export. But Tycho supports it.


> I conclude this issue thread with only some \"feature requests\":\n\nMove the feedback from the Error Log to the Eclipse Product Export Wizard or disable the export button (I'd prefer the second option regarding UX)

I think that should be implemented an the issue therefore kept open for that.

**Comment by laeubi** [comment]
I now have created an umbrella for this here:

- https://github.com/eclipse-pde/eclipse.pde/issues/862
